YDotNet.Server/Internal/DocumentWriter.cs

```diff
@@ -0,0 +1,146 @@
+
+    private sealed record WriteDocumentMessage(string Name, DocumentContainer Container) : WriteMessage(Name);
+
+    private record LastMessage
+    {
+        public IDisposable? Timer { get; set; }
+
+        public WriteMessage? Message { get; set; }
+
+        required public DateTimeOffset Created { get; init; }
+    }
+
+    public DocumentWriter(
+        IDocumentStorage storage,
+        TimeSpan delay,
+        TimeSpan delayMax,
+        ILogger logger)
+    {
+        this.storage = storage;
+
+        writeBlock = new ActionBlock<WriteMessage>(
+            async message =>
+            {
+                logger.LogDebug("Document {documentName} will be written to storage {storage}", message.Name, storage);
+
+                try
+                {
+                    byte[] state = Array.Empty<byte>();
+                    if (message is WriteDataMessage dataMessage)
+                    {
+                        state = dataMessage.State;
+                    }
+                    else if (message is WriteDocumentMessage documentMessage)
+                    {
+                        state = await documentMessage.Container.GetStateAsync().ConfigureAwait(false);
+                    }
+
+                    await storage.StoreDocAsync(message.Name, state).ConfigureAwait(false);
+
+                    logger.LogDebug("Document {documentName} has been written to storage {storage}", message.Name, storage);
+                }
+                catch (Exception ex)
+                {
+                    logger.LogError(ex, "Document {documentName} failed to write to storage {storage}", message.Name, storage);
+                }
+            },
+            dataflowBlockOptions: new ExecutionDataflowBlockOptions
+            {
+                MaxDegreeOfParallelism = 16,
+                MaxMessagesPerTask = 1,
+            })
+        {
+        };
+        delayBlock = CreateDelayBlock(writeBlock, delay, delayMax);
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        delayBlock.Complete();
+        await delayBlock.Completion.ConfigureAwait(false);
+
+        writeBlock.Complete();
+        await writeBlock.Completion.ConfigureAwait(false);
+    }
+
+    public Task WriteAsync(string name, byte[] doc)
+    {
+        return delayBlock.SendAsync((new WriteDataMessage(name, doc), false));
+    }
+
+    public Task WriteAsync(string name, DocumentContainer document)
+    {
+        return delayBlock.SendAsync((new WriteDocumentMessage(name, document), false));
+    }
+
+    private ActionBlock<(WriteMessage Message, bool Force)> CreateDelayBlock(ActionBlock<WriteMessage> target, TimeSpan delay, TimeSpan maxDelay)
+    {
+        var byName = new Dictionary<string, LastMessage>(StringComparer.Ordinal);
+
+        var scheduler = Scheduler.Default;
+
+        return new ActionBlock<(WriteMessage, bool)>(
+            async item =>
+            {
+                var (message, force) = item;
```

Comment:
how do I force the write of a document?

Reply:
it is not possible at the moment. It is used to avoid locking in the timer call. So the timer sends a message to the block again with the force flag.